util/topgen/c.py

```diff
@@ -30,27 +49,70 @@ def __init__(self, top_name, name, base_addr, size_bytes):
         self.size_bytes = size_bytes
 
     def base_addr_name(self):
-        return self.name.with_more(prefix=["top", self.top_name],
+        return self.name.with_more(prefix=self.top_name.parts,
                                    suffix=["base", "addr"])
 
     def size_bytes_name(self):
-        return self.name.with_more(prefix=["top", self.top_name],
+        return self.name.with_more(prefix=self.top_name.parts,
                                    suffix=["size", "bytes"])
 
+class CEnum(object):
+    def __init__(self, top_name, enum_name):
+        self.enum_name = Name.join(top_name, enum_name)
+        self.enum_counter = 0
+        self.finalized = False
```

Comment:
This isn't massively "RAII" (yes, I know this is Python, not C++, but still...) Maybe it would be neater for the `CEnum` constructor to take the list of constants as a list of tuples? So `plic_targets` would end up something like this:
```python
    def plic_targets(self):
        return CEnum(self._top_name,
                     Name(["plic", "target"]),
                     [(Name(["ibex", str(i)]), "Ibex Core {}".format(i))
                      for i in range(int(self.top["num_cores"]))],
                     'Final PLIC target')
```

Look, ma, no side effects! :-)

The init function would then be something like this:
```python
   def __init__(self, top_name, enum_name, enum_items, final_comment):
        self.enum_name = Name.join(top_name, enum_name)
        self.constants = []
        for idx, (const_name, const_doc) in enumerate(enum_items):
            self.constants.append((Name.join(self.enum_name, const_name),
                                   idx,
                                   const_doc))
        self.constants.append((enum_name.with_more(suffix=['last']),
                               len(self.constants),
                               r"\internal " + final_comment))

```

Obviously, I realise you'd need to do something a bit more clever to support setting values along the way (allow a dict instead of a tuple as an item?). But that can wait until you need it.

If you also replace `join` with `+` as I suggest above and moved the concatenation of `top_name` and `enum_name` up to the call site, the code becomes:

```python
   def __init__(self, enum_name, enum_items, final_comment):
        self.enum_name = enum_name
        self.constants = []
        for idx, (name, doc) in enumerate(enum_items):
            self.constants.append((enum_name + name, idx, doc))
        self.constants.append((enum_name.with_more(suffix=['last']),
                               len(self.constants),
                               r"\internal " + final_comment))

    ...

    def plic_targets(self):
        return CEnum(self._top_name.with_more(suffix=["plic", "target"]),
                     [(Name(["ibex", str(i)]), "Ibex Core {}".format(i))
                      for i in range(int(self.top["num_cores"]))],
                     'Final PLIC target')
```

And maybe the final comment should actually be the noun, so the "final" string is part of the literal in `__init__`.

Reply:
Looking at the following commit, I realise that the code in `plic_targets` gets extended a lot. But I still think it's cleaner to build up a list of things and then wrap them with a `CEnum`, rather than to make an empty `CEnum` and then add all the things.